### ISSUE TITLE
NAN in downstream detection finetuneing task

### ISSUE THREAD

**Issue by launchauto** [openai/CLIP#63]
I try to use you RN50.pt as a backbone in Faster-RCNN-coco-detection task in mmdetection arichtecture.
Firstly, I try to modified the layer name of RN50.pt  and resaved them so that the parameters can be laoded correctly.
Then, I modified ResNetV1d to match your ModifiedResNet in clip model.
During the fineturning procedure, I do not use your attention layer and just use the output of reslayer4,reslayer3,reslayer2 reslayer1 to FPN.

**Comment by launchauto** [comment]
During the training procedure the loss will be NAN right from the start

**Comment by jongwook** [comment]
Did you try #40 ?

**Comment by launchauto** [comment]
`model = dict(
    type='FasterRCNN',
    pretrained='/data1/pretrained_model/clip2r50v1d.pth',
    backbone=dict(
        type='ModifiedResNetV1d',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            scales=[8],
            ratios=[0.5, 1.0, 2.0],
            strides=[4, 8, 16, 32, 64]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[0.0, 0.0, 0.0, 0.0],
            target_stds=[1.0, 1.0, 1.0, 1.0]),
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
        loss_bbox=dict(type='L1Loss', loss_weight=1.0)),
    roi_head=dict(
        type='StandardRoIHead',
        bbox_roi_extractor=dict(
            type='SingleRoIExtractor',
            roi_layer=dict(type='RoIAlign', output_size=7, sampling_ratio=0),
            out_channels=256,
            featmap_strides=[4, 8, 16, 32]),
        bbox_head=dict(
            type='Shared2FCBBoxHead',
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=80,
            bbox_coder=dict(
                type='DeltaXYWHBBoxCoder',
                target_means=[0.0, 0.0, 0.0, 0.0],
                target_stds=[0.1, 0.1, 0.2, 0.2]),
            reg_class_agnostic=False,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='L1Loss', loss_weight=1.0))))
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            match_low_quality=True,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    rpn_proposal=dict(
        nms_across_levels=False,
        nms_pre=2000,
        nms_post=1000,
        max_num=1000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0.5,
            match_low_quality=False,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=512,
            pos_fraction=0.25,
            neg_pos_ub=-1,
            add_gt_as_proposals=True),
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=1000,
        nms_post=1000,
        max_num=1000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05,
        nms=dict(type='nms', iou_threshold=0.5),
        max_per_img=100))
dataset_type = 'CocoDataset'
data_root = '/dataset/public/coco/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1333, 800),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type='CocoDataset',
        ann_file='/dataset/public/coco/annotations/instances_train2017.json',
        img_prefix='/dataset/public/coco/images/train2017/',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
            dict(type='RandomFlip', flip_ratio=0.5),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size_divisor=32),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
        ]),
    val=dict(
        type='CocoDataset',
        ann_file='/dataset/public/coco/annotations/instances_val2017.json',
        img_prefix='/dataset/public/coco/images/val2017/',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1333, 800),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='Pad', size_divisor=32),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]),
    test=dict(
        type='CocoDataset',
        ann_file='/dataset/public/coco/annotations/instances_val2017.json',
        img_prefix='/dataset/public/coco/images/val2017/',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1333, 800),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='Pad', size_divisor=32),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]))
evaluation = dict(interval=12, metric='bbox')
optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=None)
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.001,
    step=[8, 11])
total_epochs = 12
checkpoint_config = dict(interval=1)
log_config = dict(interval=50, hooks=[dict(type='TextLoggerHook')])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
work_dir = 'work_dirs_3d/clip/'
gpu_ids = range(0, 1)
`

**Comment by launchauto** [comment]
> Did you try #40 ?

When I load your pretrained model and resave it I use 
clip_r50_model=torch.jit.load('/data1/pretrained/model/clip/RN50.pt',map_location='cpu').eval()
Then I modified some layer names and resaved it.
After build model, I add model=model.float()
It still NAN right from the start.
Does it work?

**Comment by jongwook** [comment]
The JIT code contains hard-coded dtype variables and will not work if you simply cast by calling `.float()`; this was why I had to patch those variables in [some hacky way](https://github.com/openai/CLIP/blob/main/clip/clip.py#L137-L157) to make it runnable on CPU.

Use `clip.load()` with `jit=False` to load the model in a non-JIT mode, and try again.

**Comment by launchauto** [comment]
> The JIT code contains hard-coded dtype variables and will not work if you simply cast by calling `.float()`; this was why I had to patch those variables in [some hacky way](https://github.com/openai/CLIP/blob/main/clip/clip.py#L137-L157) to make it runnable on CPU.
> 
> Use `clip.load()` with `jit=False` to load the model in a non-JIT mode, and try again.

Thx! I try jit=False and let the learning_rate to be 0.002 for 8GPU and batchsize=16(ImageNet supervision is 0.2 for batchsize=16). Hopefully this time I can get reasonable result.

**Comment by launchauto** [comment]
> > The JIT code contains hard-coded dtype variables and will not work if you simply cast by calling `.float()`; this was why I had to patch those variables in [some hacky way](https://github.com/openai/CLIP/blob/main/clip/clip.py#L137-L157) to make it runnable on CPU.
> > Use `clip.load()` with `jit=False` to load the model in a non-JIT mode, and try again.
> 
> Thx! I try jit=False and let the learning_rate to be 0.002 for 8GPU and batchsize=16(ImageNet supervision is 0.2 for batchsize=16). Hopefully this time I can get reasonable result.

This time it will not be NAN.

**Comment by launchauto** [comment]
> > > The JIT code contains hard-coded dtype variables and will not work if you simply cast by calling `.float()`; this was why I had to patch those variables in [some hacky way](https://github.com/openai/CLIP/blob/main/clip/clip.py#L137-L157) to make it runnable on CPU.
> > > Use `clip.load()` with `jit=False` to load the model in a non-JIT mode, and try again.
> > 
> > 
> > Thx! I try jit=False and let the learning_rate to be 0.002 for 8GPU and batchsize=16(ImageNet supervision is 0.2 for batchsize=16). Hopefully this time I can get reasonable result.
> 
> This time it will not be NAN.

BTW, do you have some suggestion to learning rate and optimizer when transfer to detection model. Thx~

**Comment by jongwook** [comment]
There's no universal rule on this, but I'd start with 10% the learning rate used for pre-training.

**Comment by launchauto** [comment]
> There's no universal rule on this, but I'd start with 10% the learning rate used for pre-training.

Thx. I will try.

**Comment by dragen1860** [comment]
@launchauto  Hi, What's the map when you adopt the clip backbone? is it gain improment? I would like to here some suggestion from you . Thank you.

**Comment by Edwardmark** [comment]
@launchauto Have you made any progress in detection using clip models?
